chain/chain/src/block_processing_utils.rs

```diff
@@ -0,0 +1,200 @@
+            return vec![];
+        }
+
+        let mut pool = self.0.write().expect(POISONED_LOCK_ERR);
+
+        let mut res = vec![];
+        let applied_blocks = pool.applied_blocks.drain().collect::<Vec<_>>();
+        for (block_hash, apply_result) in applied_blocks {
+            let (block, preprocess_result) = pool.preprocessed_blocks.remove(&block_hash).expect(
+                &format!("block {:?} in applied_blocks but not in preprocessed_blocks", block_hash),
+            );
+            res.push((block, preprocess_result, apply_result));
+        }
+
+        res
+    }
+
+    /// Returns whether the block has been added as preprocessed
+    pub(crate) fn is_block_preprocessed(&self, block_hash: &CryptoHash) -> bool {
+        self.0.read().expect(POISONED_LOCK_ERR).preprocessed_blocks.contains_key(block_hash)
+    }
+
+    /// Returns whether the pool is full
+    pub(crate) fn is_full(&self) -> bool {
+        self.0.read().expect(POISONED_LOCK_ERR).preprocessed_blocks.len() >= MAX_PROCESSING_BLOCKS
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::block_processing_utils::{BlockLike, BlockPreprocessInfo, BlockProcessingPool};
+    use actix::clock::{sleep, timeout};
+    use near_primitives::hash::{hash, CryptoHash};
+    use near_primitives::time::Duration;
+    use std::collections::HashSet;
+
+    #[derive(Clone)]
+    struct MockBlock {
+        hash: CryptoHash,
+    }
+    impl BlockLike for MockBlock {
+        fn hash(&self) -> &CryptoHash {
+            &self.hash
+        }
+    }
+
+    #[test]
+    // This test creates a scenario similar to how BlockProcessingPool will be used in chain.
+    // One thread adds preprocessed blocks to the pool and schedule other threads to add applied
+    // blocks to the pool. Then it waits and keeps checking if there are applied blocks in the pool
+    fn test_block_processing_pool_basic() {
+        actix::System::new().block_on(async {
+            timeout(Duration::from_secs(5), async {
+                let pool = BlockProcessingPool::new();
+                let mut hashes = HashSet::new();
+                for i in 0..10 {
+                    let hash = hash(&[i]);
+                    let res = pool.add_preprocessed_block(
+                        MockBlock { hash: hash.clone() },
+                        BlockPreprocessInfo::default(),
+                    );
+                    if res.is_ok() {
+                        hashes.insert(hash);
+                        let pool1 = pool.clone();
+                        let hash1 = hash.clone();
+                        // spwan another thread to add applied blocks later
+                        actix::spawn(async move {
+                            assert!(pool1.is_block_preprocessed(&hash1));
+                            sleep(Duration::from_secs(1)).await;
+                            pool1.add_applied_block(hash, vec![]).unwrap();
+                        });
```

Comment:
Yeah sorry the comment here is misleading. I'm aware that actix is single-threaded, and this is exactly the reason why I'm using actix instead spawning an actual thread. In the code where BlockProcessingPool will be used, where we will make do_apply_chunks async, I will also use actix::spawn instead of spawning a real thread.

Reply:
:thinking: how that would solve the problem of block processing tying up the client then? With actix::spawn, the Runtime::apply would run on the same thread that the client does, so it won't be able to process messages? 

We can make `Runtime::apply` `async` such that it can yield to async runtime, but that seems worse than just running the thing in a separate thread. 

It looks I am missing another bit of the plan here?

Reply:
Ok I think my understanding of rust futures and actix::spawn may still be wrong. My previous thoughts were, he current do_apply_chunks function uses threads in the rayon thread pool to run the actual `Runtime::apply` work https://github.com/near/nearcore/blob/master/chain/chain/src/chain.rs#L5190. So if I make this function async, and use actix::spawn to spawn the future in ClientActor, the client actor thread won't be blocked. But now I see that `do_apply_chunks` won't yield when it's waiting for the rayon threads to do the actual work, or will it?

Reply:
Yeah, it won't yield, only explicit awaits would. But it should be easy to move the whole thing to a rayon's pool with https://docs.rs/rayon/latest/rayon/fn.spawn.html